### ISSUE TITLE
Path Data cropping support

### ISSUE THREAD

**Issue by zero41120** [openseadragon/openseadragon#1778]
Base on my implementation for #1540 it only supports polygon.
https://github.com/openseadragon/openseadragon/pull/1772

I want to extend the coverage to support Path Data, which can be done by providing Path2D object in the [canvas clip function](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/clip#Specifying_a_path_and_a_fillRule)

However, [Path2D ](https://developer.mozilla.org/en-US/docs/Web/API/Path2D/Path2D#Browser_compatibility) is not supported on IE11

It doesn't seem like OSD had any dependency before
![image](https://user-images.githubusercontent.com/10471270/74688360-25d1d000-518c-11ea-89af-e6eae9d58084.png)

@iangilman What will be your recommendation of adding dependency?

**Comment by iangilman** [comment]
OpenSeadragon currently has no dependencies, which I think is good. I think it would be best not to add the path2d polyfill to OpenSeadragon. Perhaps we could add the path data cropping support and document that it requires this polyfill if you want it on IE11 (and gracefully degrade if it's not present).

Adding Babel would be a major change... I'm hesitant to do so, but perhaps we should discuss the pros and cons. That should happen in a separate issue just for the topic.

**Comment by msalsbery** [comment]
@zero41120 Perhaps you could enhance the work on #1540 to make it hookable/extendible by plugins. Then you can use whatever libraries and build tools you’d like on the plugins

**Comment by zero41120** [comment]
@msalsbery 
I was attempting to do a plugin before I made my implementation viable to public.
The issue is that drawTiles is a private function which plugin cannot access it from the OpenSeadragon object.

**Comment by zero41120** [comment]
@iangilman 
I agree with notifying user to include poly-fill if they want to use IE.

I found it really difficult to translate the path object to match in OSD's viewport because paths are not points. We cannot just use the existing code to convert them to drawer coordinate if the path data uses arc and other things.

**Comment by iangilman** [comment]
> The issue is that drawTiles is a private function which plugin cannot access it from the OpenSeadragon object.

Yeah, unfortunately we have some private functions like that that make it hard to customize some areas. I think the solution is to make those functions not private! I'd be happy to have a PR to that effect.

> I found it really difficult to translate the path object to match in OSD's viewport because paths are not points. We cannot just use the existing code to convert them to drawer coordinate if the path data uses arc and other things.

Indeed... I guess we'd need to use the canvas transform methods instead, to do the conversion in this case.

**Comment by zero41120** [comment]
> Indeed... I guess we'd need to use the canvas transform methods instead, to do the conversion in this case.
Do you have any suggestion for the transform? I don't know much about this topic.

**Comment by iangilman** [comment]
I think basically it means redoing this section: 

https://github.com/openseadragon/openseadragon/blob/3adb001b9f6a5da98bb4a7056e0063c737207d21/src/tiledimage.js#L1993-L1997

...in terms of canvas transforms. So, for instance, you'd start with `imageToViewportCoordinates`, which calls `_imageToViewportDelta`, which contains this code:

```
var scale = this._contentBoundsNoRotate.width;
return new $.Point(
    imageX / this._contentSizeNoRotate.x * scale,
    imageY / this._contentSizeNoRotate.x * scale
);
```

That might translate into something like (if I'm interpreting it properly):

```
var scale = 1 / (this._contentSizeNoRotate.x * this._contentBoundsNoRotate.width);
context.scale(scale);
```

It (imageToViewportCoordinates) then has this code:

```
point.x += this._contentBoundsNoRotate.x;
point.y += this._contentBoundsNoRotate.y;
```

...which might become:

```
context.translate(this._contentBoundsNoRotate.x, this._contentBoundsNoRotate.y);
```

And you would keep going through the code, adding scales, translates, and rotates as you go. Also, you should start the whole thing off like so:

```
var oldMatrix = context.getTransform();
```

...and end it up after you clip it with:

```
context.setTransform(oldMatrix);
```

...so you're not transforming everything else afterwards.

So, I think it's definitely doable, but it will take a bit of spelunking through the code, and probably some trial and error to get the stack of transforms proper.

**Comment by zero41120** [comment]
I don't really understand canvas transform, so I have not attempt to provide such change to OSD.
But I made this approximation function to convert path to polygon
```javascript
let pathToPolygon = (path, segments = 360) => {
  const pathLength = path.getTotalLength();
  const points = [];
  
  for (let i = 0; i < segments; i++) {
    const trace = i * pathLength / (segments-1);
    const {x, y} = path.getPointAtLength(trace);
    points.push({x, y}); // You can replace this with OpenSeadragon.Point
  }
  return points;
}
```

Given a path's `d` string, you can convert that to path object that the above function uses:
```javascript
let dToPath = (d) => {
  const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
  path.setAttribute('d', d);
  return path;
}
```

**Comment by iangilman** [comment]
Cool. So this provides at least an approximate way to support arbitrary paths via the existing setCroppingPolygons mechanism? Are you proposing we add this directly to OpenSeadragon, or leave it as something people can call from outside?

**Comment by zero41120** [comment]
@iangilman 
We can probably add that and explicitly emphasize it is an approximation.
There might be 2 problems if the user doesn't know what they are doing.
But it really depends on the use case. 
- With the default configuration (360 segments), a small square of 4 points will generate 360 segments that are mostly useless. If they use a ton of square path, performance may suffer.
- Also with default, a large arcs may not be enough curvature in the red areas since my code just evenly divided the length of the path. If they use too few segments, precision may suffer.
![image](https://user-images.githubusercontent.com/10471270/84885159-817f6d80-b047-11ea-9880-23355f4380cf.png)

**Comment by iangilman** [comment]
Indeed. Well, given these limitations, and the fact that it doesn't require tight integration with OpenSeadragon, I think it would be better as a plug-in or even example code (in this collection for instance: https://codepen.io/collection/APoyjJ/).

One thought is you could generate a ton of points and then use https://mourner.github.io/simplify-js/ to reduce them in a smart fashion.

**Comment by zero41120** [comment]
@iangilman One more issue I have identified.
My approximation does pretty okay job in most cases, but if the path is not one continuous path, then there will be some gaps.
Say your path consist of a box and a shape inside it as exclusion:
![image](https://user-images.githubusercontent.com/10471270/89220884-3255b380-d587-11ea-88d4-30510ff7a9c2.png)

My approximation as polygon will try to connect the outer box to the inner exclusion,
thus create gap similar to this.
![image](https://user-images.githubusercontent.com/10471270/89221138-a001df80-d587-11ea-96c9-4e8471dd3af6.png)

**Comment by iangilman** [comment]
Good point. Well, if you do a plug-in or other such library, you can document the weak sides of the algorithm and work to fix them when you can.

**Comment by zero41120** [comment]
@iangilman 
I attempted to add the path cropping feature, here is the compare
https://github.com/openseadragon/openseadragon/compare/master...zero41120:feature/1778-path-cropping

It doesn't seem like the suggestion for transformation and scale works, can you elaborate more on this?

**Comment by zero41120** [comment]
Pull request created here: 
https://github.com/openseadragon/openseadragon/pull/1873
